File: Leetcode/42-Trapping-Rain-Water/TrappingRainWater.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        total = 0
        if not height or len(height)<3:
            return total

        length = len(height)
        left = [0]*length
        right = [0]*length

        m = height[0]
        for i in range(1, length):
            left[i] = max(m, height[i-1])
            m = max(m, height[i])

        m = height[length-1]
        for i in range(length-2, 0, -1):
            right[i] = max(m, height[i+1])
            m = max(m, height[i])

        s = 0
        for i in range(1, length-1):
            s = min(left[i], right[i])-height[i]
            if s>0:
                total += s
        return total
    

File: Lintcode/363-Trapping-Rain-Water/TrappingRainWater.java

public class TrappingRainWater {
	/**
     * @param heights: an array of integers
     * @return: a integer
     */
	public int trapRainWater(int[] heights) {
		if (heights == null || heights.length < 3) {
			return 0;
		}
		int n = heights.length;
		int[] left = new int[n];
		int[] right = new int[n];

		int max = heights[0];
		for (int i = 0; i < n; i++) {
			max = Math.max(max, heights[i]);
			left[i] = Math.max(max, heights[i]);
		}

		max = heights[n - 1];
		for (int i = n - 1; i >= 0; i--) {
			max = Math.max(max, heights[i]);
			right[i] = Math.max(max, heights[i]);
		}

		max = 0;
		for (int i = 1; i < n - 1; i++) {
			max += Math.min(left[i], right[i]) - heights[i];
		}
		return max;
	}
}
